src/stdlib_selection.fypp

```diff
@@ -0,0 +1,275 @@
+! (https://github.com/fortran-lang/stdlib/pull/500#commitcomment-57418593)
+!
+
+use stdlib_kinds
+
+implicit none
+
+private
+
+public select, arg_select
+
+interface select
+    !! version: experimental
+    !! ([Specification](..//page/specs/stdlib_selection.html#select-find-the-k-th-smallest-value-in-an-input-array))
+
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("select", 1, arraytype, arraykind, intkind)
+      module procedure ${name}$
+    #:endfor
+  #:endfor
+end interface
+
+interface arg_select
+    !! version: experimental
+    !! ([Specification](..//page/specs/stdlib_selection.html#arg_select-find-the-index-of-the-k-th-smallest-value-in-an-input-array))
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("arg_select", 1, arraytype, arraykind, intkind)
+      module procedure ${name}$
+    #:endfor
+  #:endfor
+end interface
+
+contains
+
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("select", 1, arraytype, arraykind, intkind)
+      subroutine ${name}$(a, k, kth_smallest, left, right)
+          !! select - select the k-th smallest entry in a(:).
+          !!
+          !! Partly derived from the "Coretran" implementation of 
+          !! quickSelect by Leon Foks, https://github.com/leonfoks/coretran
+          !!
+          ${arraytype}$, intent(inout) :: a(:)
+              !! Array in which we seek the k-th smallest entry.
+              !! On output it will be partially sorted such that
+              !! `all(a(1:(k-1)) <= a(k)) .and. all(a(k) <= a((k+1):size(a)))`.
+          ${inttype}$, intent(in) :: k
+              !! We want the k-th smallest entry. E.G. `k=1` leads to
+              !! `kth_smallest=min(a)`, and `k=size(a)` leads to
+              !! `kth_smallest=max(a)`
+          ${arraytype}$, intent(out) :: kth_smallest
+              !! On output contains the k-th smallest value of `a(:)`
+          ${inttype}$, intent(in), optional :: left, right
+              !! If we know that:
+              !!    the k-th smallest entry of `a` is in `a(left:right)`
+              !! and also that:
+              !!    `maxval(a(1:(left-1))) <= minval(a(left:right))`
+              !! and:
+              !!    `maxval(a(left:right))) <= minval(a((right+1):size(a)))`
+              !! then one or both bounds can be specified to narrow the search.
+              !! The constraints are available if we have previously called the
+              !! subroutine with different `k` (because of how `a(:)` becomes
+              !! partially sorted, see documentation for `a(:)`).
+
+          ${inttype}$ :: l, r, mid, iPivot
+          integer, parameter :: ip = ${intkind}$
+
+          l = 1_ip
+          if(present(left)) l = left
+          r = size(a, kind=ip)
+          if(present(right)) r = right
```

Comment:
Consider simplifying with `optval`. It seems like the overhead from `optval` would be negligible relative to the work done in the rest of the subroutine.

Reply:
I've opened a thread to discuss this point: https://github.com/fortran-lang/stdlib/issues/574

Reply:
To expedite merging this PR, let's not worry about `optval` for now, since it's only a stylistic change and not a functional one. We can easily address it in a future PR if we decide we want it.